project_generator/project.py

```diff
@@ -0,0 +1,375 @@
+                self._process_source_files([os.path.join(source_file, f) for f in os.listdir(
+                    source_file) if os.path.isfile(os.path.join(source_file, f))], group_name)
+
+            extension = source_file.split('.')[-1]
+            extension = mappings[extension] or extension
+
+            if extension not in extensions:
+                continue
+
+            if extension not in self.source_groups[group_name]:
+                self.source_groups[group_name][extension] = []
+
+            self.source_groups[group_name][extension].append(source_file)
+
+            if os.path.dirname(source_file) not in self.source_paths:
+                self.source_paths.append(os.path.dirname(source_file))
+
+    def clean(self, tool):
+        if tool is None:
+            tools = list(self.TOOLCHAINS)
+        else:
+            tools = [tool]
+
+        for current_tool in tools:
+            path = os.path.join(self.project_dir['path'], "%s_%s" % (current_tool, self.name))
+
+            if os.path.isdir(path):
+                logging.info("Cleaning directory %s" % path)
+
+                shutil.rmtree(path)
+
+
+    def build(self, tool):
+        """build the project"""
+        build(self.project_path, self.project_files, tool, self.workspace.settings)
+
+    def export(self, tool):
+        """export the project"""
+        project_path, project_files = export(
+            self.generate_dict_for_tool(tool), tool, self.workspace.settings)
+
+        self.project_path = project_path
+        self.project_files = project_files
+
+        return project_path, project_files
+
+    def source_of_type(self, filetype):
+        """return a dictionary of groups and the sources of a specified type within them"""
+        files = {}
+        for group, group_contents in self.source_groups.items():
+            files[group] = []
+            if filetype in group_contents:
+                files[group].extend(group_contents[filetype])
+
+        return files
+
+    def all_sources_of_type(self, filetype):
+        """return a list of the sources of a specified type"""
+        files = []
+
+        for group, group_contents in self.source_groups.items():
+            if filetype in group_contents:
+                files.extend(group_contents[filetype])
+
+        return files
+
+    def generate_dict_for_tool(self, tool):
+        """for backwards compatibility"""
+        tool_specific_settings = self.tool_specific[self.TOOLCHAINS[tool]]
+        d = {
+            'name': self.name,
+            'mcu': self.mcu,
+            'core': self.core,
+            'target': self.target,
+            'include_paths': self.include_paths + tool_specific_settings.include_paths,
+            'source_paths': self.source_paths + tool_specific_settings.source_paths,
+            'source_files': merge_recursive(self.source_groups, tool_specific_settings.source_groups),
+            # for backwards compatibility
```

Comment:
basically I don't really mean backwards compatibility I mean this so that the data is in the same format as the old ProjectGenerator